aws/rust-runtime/aws-auth/src/refresh.rs

```diff
@@ -0,0 +1,294 @@
+const REFRESH_TIMEOUT: Duration = Duration::from_secs(5);
+const REFRESH_BUFFER_TIME: Duration = Duration::from_secs(30);
+const MAX_WAIT_TIME: Duration = Duration::from_millis(1000);
+
+const WATCHDOG_STACK_SIZE_BYTES: usize = 256;
+
+pub type AsyncCredentialResult = Result<Credentials, CredentialsError>;
+
+// TODO: Add doc comment
+pub trait AsyncCredentialLoader: Send + Sync {
+    fn load_credentials(&self) -> Pin<Box<dyn Future<Output = AsyncCredentialResult> + Send>>;
+}
+
+mod error {
+    use std::error::Error as StdError;
+    use std::fmt;
+
+    #[derive(Debug)]
+    #[non_exhaustive]
+    pub enum Error {
+        RefreshFailed(Box<dyn StdError + Send + Sync>),
+        ThreadSpawnFailed(Box<dyn StdError + Send + Sync>),
+        ExecutorThreadFailed,
+    }
+
+    impl StdError for Error {}
+
+    impl fmt::Display for Error {
+        fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+            match self {
+                Error::RefreshFailed(err) => {
+                    write!(f, "credential refresh failed: {}", err)
+                }
+                Error::ThreadSpawnFailed(err) => {
+                    write!(f, "failed to spawn thread: {}", err)
+                }
+                Error::ExecutorThreadFailed => write!(f, "executor thread panicked"),
+            }
+        }
+    }
+}
+
+pub use error::Error;
+
+// TODO: Consider using https://docs.rs/arc-swap/1.3.0/arc_swap/
+/// Stores a value in a RwLock and never holds the lock open across a section
+/// of code that can panic in order to eliminate the PoisonedError.
+#[derive(Clone)]
+struct RwCell<T: Clone> {
+    value: Arc<RwLock<T>>,
+}
+
+impl<T: Clone> RwCell<T> {
+    pub fn new(initial: T) -> RwCell<T> {
+        RwCell {
+            value: Arc::new(RwLock::new(initial)),
+        }
+    }
+
+    pub fn get(&self) -> T {
+        self.value.read().expect("cannot be poisoned").clone()
+    }
+
+    pub fn set(&self, value: T) {
+        *self.value.write().expect("cannot be poisoned") = value;
+    }
+}
+
+#[derive(Clone)]
+struct Inner {
+    loader: Arc<dyn AsyncCredentialLoader>,
+    credentials: RwCell<Option<Credentials>>,
+}
+
+impl Inner {
+    pub async fn refresh(&self) -> Result<(), CredentialsError> {
+        let credentials = self.loader.load_credentials().await?;
+        self.credentials.set(Some(credentials));
+        Ok(())
+    }
+
+    fn calc_next_refresh_sleep_time(now: SystemTime, expiration: SystemTime) -> Option<Duration> {
+        let expiration = expiration - REFRESH_BUFFER_TIME;
+        if now >= expiration {
+            None
+        } else {
+            let remaining = expiration.duration_since(now).expect("now < expiration");
+            if remaining > MAX_WAIT_TIME {
+                Some(MAX_WAIT_TIME)
+            } else {
+                Some(remaining)
+            }
+        }
+    }
+
+    fn next_refresh_sleep_time(&self, now: SystemTime) -> Option<Duration> {
```

Comment:
it's not immediately obvious what `None` means here, I might consider introducing an enum like:

```rust
enum Refresh {
  After(Duration),
  Now
}
```

to make it totally clear